### ISSUE TITLE
Can I call uc_emu_start() and uc_emu_stop() from different threads?

### ISSUE THREAD

**Issue by xorstream** [unicorn-engine/unicorn#292]
Should I be able to start the emu in one thread by doing uc_emu_start(), and then stop it from another thread by doing uc_emu_stop()? My aim is to have the emulation running in the background and be able to stop it from another thread.

When I attempt to do so in windows I get an error when I call uc_emu_start() which I think is due to it being called from a different thread to the one that I called uc_emu_open() from.

Error msg:
Assertion failed!

Program: C:\unicorn_msvc\dyn\Debug\dyn.exe
File: util/qemu-thread-win32.c, Line 70

Expression: mutex->owner == GetCurrentThreadId()

This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.

This points to the following assertion:
    void qemu_mutex_unlock(QemuMutex *mutex)
    {
        assert(mutex->owner == GetCurrentThreadId());
        mutex->owner = 0;
        LeaveCriticalSection(&mutex->lock);
    }

So I am wondering if this behaviour is by design? And if so what is the expected way to run the emu in the background?

If a thread is busy running uc_emu_start() then it cannot do its own calls to uc_emu_stop(), except for in the callback hooks or if uc_emu_start() had a count or timeout limit. And so if other threads also cannot call uc_emu_stop() then I guess I have to do something like check a flag after a timeout or in code hooks for if uc_emu_stop() should be called.

**Comment by farmdve** [comment]
I was just about to come here to ask about thread-safety myself.

I basically wish to have a GUI thread that displays registers, maybe even memory, but I am not too sure how thread-safe Unicorn is, can I start/stop from a different thread? Is reading or writing to registers thread-safe?

**Comment by aquynh** [comment]
Unicorn is thread-safe by design so yes you can stop emulation from other thread. if you find something wrong in implementation, please open an issue with details on how to reproduce it. it is best to do this with a pull request to put the sample code under tests/ directory.

**Comment by xorstream** [comment]
The problem appears to be due to the mutex implementation in "unicorn\qemu\util\qemu-thread-win32.c". It uses critical sections for it's mutexes, however critical sections can only be used within a single thread. So when you try to do stuff from other (multiple) threads it fails.

The fix is to change it from using critical sections to actual mutexes/semaphores. I can make this change, but atm I am battling against git hehe. Once I work out how to bend git to my will I will make some fixes for it.

**Comment by univm** [comment]
CriticalSections are probably used for faster lock/unlock vs mutexes.

**Comment by xorstream** [comment]
Yes critical sections can provide fast locks, but they only work within one thread. So you cannot lock from one thread and then unlock from another. Since the mutexes here are supposed to synchronise access across multiple threads, using critical sections is not a good choice :)

qemu_mutex_lock() / qemu_mutex_unlock() are called by qemu_mutex_lock_iothread() / qemu_mutex_unlock_iothread() which come with the info:

/**
 * qemu_mutex_lock_iothread: Lock the main loop mutex.
 *
 * This function locks the main loop mutex.  The mutex is taken by
 * qemu_init_main_loop and always taken except while waiting on
 * external events (such as with select).  The mutex should be taken
 * by threads other than the main loop thread when calling
 * qemu_bh_new(), qemu_set_fd_handler() and basically all other
 * functions documented in this file.
 *
 * NOTE: tools currently are single-threaded and qemu_mutex_lock_iothread
 * is a no-op there.
 */
void qemu_mutex_lock_iothread(struct uc_struct* uc);

/**
 * qemu_mutex_unlock_iothread: Unlock the main loop mutex.
 *
 * This function unlocks the main loop mutex.  The mutex is taken by
 * qemu_init_main_loop and always taken except while waiting on
 * external events (such as with select).  The mutex should be unlocked
 * as soon as possible by threads other than the main loop thread,
 * because it prevents the main loop from processing callbacks,
 * including timers and bottom halves.
 *
 * NOTE: tools currently are single-threaded and qemu_mutex_unlock_iothread
 * is a no-op there.
 */
void qemu_mutex_unlock_iothread(struct uc_struct* uc);


I don't know what the "tools" are that this refers to, but it definitely mentions that the mutexes are expected to work across multiple threads. The callstack for uc_emu_start(), which fails when called from another thread in the windows implementation, is:

uc_emu_start() (uc.c)
vm_start() (qemu\cpus.c)
qemu_mutex_unlock_iothread() (qemu\mainloop.c)
qemu_mutex_unlock() (qemu\util\qemu-thread-win32.c)

qemu_mutex_unlock() throws an assert error due to:
assert(mutex->owner == GetCurrentThreadId());

**Comment by univm** [comment]
As I am sure you are aware, Unicorn is a stripped-down version of original QEMU, thus perhaps the design choices made with the intent of full HW virtualization no longer apply, or not as much.

**Comment by xorstream** [comment]
Yes I am looking at this purely from a unicorn point of view. The files are inside the "qemu" dir, but I am only looking at what they are doing.

qemu\util\qemu-thread-win32.c is the windows version of qemu\util\qemu-thread-posix.c.
The qemu-thread-posix.c implementation uses real mutexes that work across threads.

Without using real mutexes, ie ones that work across threads, you cannot call uc_emu_start() from another thread. This means that you *cannot* run the emulator in the background while using a separate thread to control it, such as a user interface. This was the subject of my first post in the issue, and one that was answered by @aquynh as being something that unicorn is supposed to support.

**Comment by lunixbochs** [comment]
I can confirm this behavior (which is why I filed #170). Unicorn will sometimes abort in a mutex function if I change the host thread from which I'm calling the API. I worked around this by forcing Usercorn onto one OS thread, which is why I haven't followed up yet.

**Comment by JonathonReinhart** [comment]
Note also, that unicorn hooks are called from a different thread than the
one from which you call uc_emu_start(). They're called from a qemu vcpu
thread.

**Comment by xorstream** [comment]
Ok so after lots of trying stuff and reading the important things are.

* Critical sections *can* be used across multiple threads, they are just limited to being inside one process. (i thought it was a one thread limit).

* The current win32 implementation of mutexes actually does work just like the pthread implementation. This is that recursive mutexes are not allowed and a mutex must be unlocked by the same thread that locked it.

So the upshot of this is that the mutex locking operation is the same for win32 and posix, so its best to leave it as it is. The issue then is that calling uc_emu_start() from a separate thread will not work currently due to how it uses mutexes. So now my plan is to look into how it use mutexes and hopefully change the way it is done somewhat to make this work.

(I also came across a few crashes due to concurrency issues while playing around, but that's another story)

**Comment by JonathonReinhart** [comment]
There are plenty of APIs that have the same multithreading constraints as Unicorn. Why can't you change your application to call Unicorn from the same thread?

**Comment by xorstream** [comment]
Main reason is that I want to run the emu in the background while other stuff is done in the foreground.

uc_emu_start() blocks until completion, so unless you continually use timeout out or run 1 instruction at a time, then check for other unicorn related events, you can't do anything else in the same thread.

I've worked out now that the issue about calling start from another thread is that uc_open() locks uc->qemu_global_mutex and then uc_emu_start() unlocks it. Due to the mutex implementation these functions both need to be called from the same thread. This is not obvious or mentioned anywhere, and from my point of view is an unnecessary limitation.

**Comment by JonathonReinhart** [comment]
I see. Yes, there are definitely threading issues that are not obvious or documented.

**Comment by lunixbochs** [comment]
I've even had success with critical sections using mmap'd shared memory across processes. I'm inclined to believe this is a logic error.

**Comment by lunixbochs** [comment]
is this still a problem with no-thread? the iothread lock should be gone now

**Comment by xorstream** [comment]
This appears to be fixed now.
I just ran the test "threaded_emu_start" that I had added back when I was having the issue, and it appears to complete successfully.
